crates/sui-network/src/randomness/mod.rs

```diff
@@ -0,0 +1,686 @@
+        };
+
+        let sig_bounds = (
+            Bound::Included((epoch, round, PeerId([0; 32]))),
+            Bound::Excluded((epoch, round + 1, PeerId([0; 32]))),
+        );
+
+        // If we have enough partial signatures, aggregate them.
+        let sig_range = self
+            .received_partial_sigs
+            .range(sig_bounds)
+            .flat_map(|(_, sigs)| sigs);
+        let mut sig =
+            match ThresholdBls12381MinSig::aggregate(self.aggregation_threshold, sig_range) {
+                Ok(sig) => sig,
+                Err(fastcrypto::error::FastCryptoError::NotEnoughInputs) => return, // wait for more input
+                Err(e) => {
+                    error!("error while aggregating randomness partial signatures: {e:?}");
+                    return;
+                }
+            };
+
+        // Try to verify the aggregated signature all at once. (Should work in the happy path.)
+        if ThresholdBls12381MinSig::verify(vss_pk.c0(), &round.signature_message(), &sig).is_err() {
+            // If verifiation fails, some of the inputs must be invalid. We have to go through
+            // one-by-one to find which.
+            // TODO: add test for individual sig verification.
+            self.received_partial_sigs
+                .retain(|&(e, r, peer_id), partial_sigs| {
+                    if epoch != e || round != r {
+                        return true;
+                    }
+                    if ThresholdBls12381MinSig::partial_verify_batch(
+                        vss_pk,
+                        &round.signature_message(),
+                        partial_sigs.iter(),
+                        &mut rand::thread_rng(),
+                    )
+                    .is_err()
+                    {
+                        warn!(
+                            "received invalid partial signatures from possibly-Byzantine peer {peer_id}"
+                        );
+                        // TODO: Ignore future messages from peers sending bad signatures.
+                        return false;
+                    }
+                    true
+                });
+            let sig_range = self
+                .received_partial_sigs
+                .range(sig_bounds)
+                .flat_map(|(_, sigs)| sigs);
+            sig = match ThresholdBls12381MinSig::aggregate(self.aggregation_threshold, sig_range) {
+                Ok(sig) => sig,
+                Err(fastcrypto::error::FastCryptoError::NotEnoughInputs) => return, // wait for more input
+                Err(e) => {
+                    error!("error while aggregating randomness partial signatures: {e:?}");
+                    return;
+                }
+            };
+            if let Err(e) =
+                ThresholdBls12381MinSig::verify(vss_pk.c0(), &round.signature_message(), &sig)
+            {
+                error!("error while verifying randomness partial signatures after removing invalid partials: {e:?}");
```

Comment:
this should be impossible, right? at least warrants a debug assert

Reply:
Added